psiminer-core/src/main/kotlin/psi/Utils.kt

```diff
@@ -53,3 +57,31 @@ fun PsiElement.printTree(delimiter: String = "..", indentStep: Int = 2) {
         println(representation.toString())
     }
 }
+
+@Serializable
+data class Position(
+    @SerialName("l") val line: Int,
+    @SerialName("c") val column: Int
+)
+
+@Serializable
+data class NodeRange(
+    val start: Position,
+    val end: Position
+)
+
+class PositionConverter(private val document: Document?) {
+    constructor(root: PsiElement) : this(PsiDocumentManager.getInstance(root.project).getDocument(root.containingFile))
+
+    fun generateNodeRange(el: PsiElement): NodeRange? {
+        if (document == null) return null
+        val startLine = document.getLineNumber(el.textRange.startOffset)
+        val endLine = document.getLineNumber(el.textRange.endOffset)
+        val startColumn = el.textRange.startOffset - document.getLineStartOffset(startLine)
+        val endColumn = el.textRange.endOffset - document.getLineStartOffset(endLine)
+        return NodeRange(
+            Position(startLine + 1, startColumn + 1),
+            Position(endLine + 1, endColumn + 1)
+        )
+    }
+}
```

Comment:
Maybe extracting position as extension function would be more readable?
```kotlin
val position = root.positionRange()
```

And I think making class is not a good idea. Storage will create a new instance for each tree that I found kind of useless. I suggest to always extract documents through managers:
```kotlin
@Serializable
data class Position(@SerialName("l") val line: Int, @SerialName("c") val column: Int) {
    constructor(document: Document, offset: Int) : this(
        document.getLineNumber(offset),
        offset - document.getLineStartOffset(document.getLineNumber(offset))
    )
}
fun PsiElement.nodeRange() {
    val document = PsiDocumentManager.getInstance(project).getDocument(containingFile)
    val textRange = getTextRange()
    val startPosition = Position(document, textRange.startOffset)
    val endPosition = Position(documents, textRange.endOffset)
    return NodeRange(startPosition, endPosition)
}
```

Reply:
Is it okay to call `PsiDocumentManaget.getInstance().getDocument()` on every PsiElement in every tree rather than creating new `PositionConverter` for every tree root?

Reply:
I think that should be okay, AFAIK, they use something like singleton for these getters.
But we can parameterize `nodeRange` with `Document` instance to reduce the number of such calls.